source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -569,21 +569,17 @@ void AggressiveDCEPass::InitializeModuleScopeLiveInstructions() {
   }
   // Keep all entry points.
   for (auto& entry : get_module()->entry_points()) {
-    if (get_module()->version() >= SPV_SPIRV_VERSION_WORD(1, 4) &&
-        !preserve_interface_) {
-      // In SPIR-V 1.4 and later, entry points must list all global variables
-      // used. DCE can still remove non-input/output variables and update the
-      // interface list. Mark the entry point as live and inputs and outputs as
-      // live, but defer decisions all other interfaces.
+    if (!preserve_interface_) {
       live_insts_.Set(entry.unique_id());
       // The actual function is live always.
       AddToWorklist(
           get_def_use_mgr()->GetDef(entry.GetSingleWordInOperand(1u)));
       for (uint32_t i = 3; i < entry.NumInOperands(); ++i) {
         auto* var = get_def_use_mgr()->GetDef(entry.GetSingleWordInOperand(i));
         auto storage_class = var->GetSingleWordInOperand(0u);
-        if (storage_class == SpvStorageClassInput ||
-            storage_class == SpvStorageClassOutput) {
+        // Vulkan support outputs without an associated input, but not inputs
+        // without an associated output.
+        if (storage_class == SpvStorageClassOutput) {
```

Comment:
Should this still be protected against old spirv versions?  Or has this always been like this?

Reply:
No, now all interfaces except output can be removed if they're dead. Previously we always kept input too due to ambiguity in the Vulkan spec.

Reply:
Oh, nice.  Thanks.

Reply:
Oh, nice.  Thanks.

I think some of the legalization checks are getting unexpected DCE activity that's messing up the pattern matching (e.g., `ToSpirv/HlslLegalizeTest.FromFile/hlsl_flattenSubset_frag`)

Reply:
I'll have to fix that when this is rolled into glslang. They look harmless.

Reply:
The only reason for the spir-v version check was that entry points in earlier versions of spir-v would not have anything to remove because everything on the entry point was either an input or output variable.  With this clarification it is safe to remove the version check.

Reply:
Vulkan may not but there are plenty of transpiled cases to MSL/HLSL where this totally broke our shader codegen.  This should have had an option to enable the old behavior.